File: mirror/src/ml/normalize.js
// Port of ml/src/normalize.py — must stay numerically identical to it, since
// the classifier was trained on vectors produced by that script. Re-centers on
// the wrist and scales by hand span so recognition is invariant to distance/
// position from the camera (see root CLAUDE.md — not optional polish).

const WRIST_IDX = 0;
const MIDDLE_MCP_IDX = 9;
const NUM_LANDMARKS = 21;
const VEC_LEN_PER_HAND = NUM_LANDMARKS * 3;

// landmarks: array of 21 {x, y, z} (e.g. MediaPipe Hands output), or null/
// undefined if that hand wasn't detected. Returns a flat length-63 Float32Array.
export function normalizeHand(landmarks) {
  if (!landmarks) {
    return new Float32Array(VEC_LEN_PER_HAND);
  }

  const wrist = landmarks[WRIST_IDX];
  const centered = landmarks.map((p) => [p.x - wrist.x, p.y - wrist.y, p.z - wrist.z]);

  const mcp = centered[MIDDLE_MCP_IDX];
  const handSpan = Math.sqrt(mcp[0] * mcp[0] + mcp[1] * mcp[1] + mcp[2] * mcp[2]);
  const scale = handSpan > 1e-6 ? handSpan : 1.0;

  const out = new Float32Array(VEC_LEN_PER_HAND);
  for (let i = 0; i < centered.length; i++) {
    out[i * 3] = centered[i][0] / scale;
    out[i * 3 + 1] = centered[i][1] / scale;
    out[i * 3 + 2] = centered[i][2] / scale;
  }
  return out;
}

// Fixed-shape (126,) vector: 63 for left hand + 63 for right hand, in that
// order. Don't change the shape or ordering without updating ml/ and CLAUDE.md.
export function landmarksToVector(leftHandLandmarks, rightHandLandmarks) {
  const left = normalizeHand(leftHandLandmarks);
  const right = normalizeHand(rightHandLandmarks);
  const out = new Float32Array(left.length + right.length);
  out.set(left, 0);
  out.set(right, left.length);
  return out;
}

File: ml/src/normalize.py
"""Landmark normalization shared by extraction (offline) and live inference (mirror app
port of this logic). Re-centers on the wrist and scales by hand span so recognition is
invariant to distance/position from the camera — see root CLAUDE.md, this step is not
optional.
"""

import numpy as np

WRIST_IDX = 0
MIDDLE_MCP_IDX = 9
NUM_LANDMARKS = 21
VEC_LEN_PER_HAND = NUM_LANDMARKS * 3  # x, y, z


def normalize_hand(landmarks_xyz):
    """landmarks_xyz: (21, 3) array for one hand, or None if not detected.
    Returns a flat (63,) float32 vector, zeros if the hand wasn't detected."""
    if landmarks_xyz is None:
        return np.zeros(VEC_LEN_PER_HAND, dtype=np.float32)

    pts = np.asarray(landmarks_xyz, dtype=np.float32)
    wrist = pts[WRIST_IDX]
    centered = pts - wrist

    hand_span = np.linalg.norm(centered[MIDDLE_MCP_IDX])
    scale = hand_span if hand_span > 1e-6 else 1.0
    scaled = centered / scale

    return scaled.flatten()


def landmarks_to_vector(left_hand_xyz, right_hand_xyz):
    """Fixed-shape (126,) vector: 63 for left hand + 63 for right hand, in that
    order. This is the contract's "fixed-shape landmark array" — don't change the
    shape or ordering without updating mirror/ and CLAUDE.md."""
    left = normalize_hand(left_hand_xyz)
    right = normalize_hand(right_hand_xyz)
    return np.concatenate([left, right])
